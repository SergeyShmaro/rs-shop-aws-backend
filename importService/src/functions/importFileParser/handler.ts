import type { S3Handler } from 'aws-lambda';
import csv from 'csv-parser';
import type { GetObjectCommandOutput } from '@aws-sdk/client-s3';
import { SQSClient, SendMessageCommand } from '@aws-sdk/client-sqs';
import { S3Client, GetObjectCommand, CopyObjectCommand, DeleteObjectCommand } from '@aws-sdk/client-s3';
import { BUCKET_NAME, REGION } from 'src/constants/environment';

const getFileName = (key: string) => {
  const splittedKey = key.split('/');
  const fileNameWithExtension = splittedKey[splittedKey.length - 1];
  if (fileNameWithExtension.indexOf('.csv') === -1) return undefined;
  return fileNameWithExtension;
};

const sendDataToQueue = (sqsClient: SQSClient, data: any) => {
  if (typeof data !== 'object') return;
  if (typeof data.title !== 'string') return;
  if (data.imageSrc !== undefined && typeof data.imageSrc !== 'string') return false;
  if (data.description !== undefined && typeof data.description !== 'string') return false;
  const price = typeof data.price === 'string' ? parseFloat(data.price) : data.price;
  if (typeof price !== 'number' || price <= 0) return;
  const count = typeof data.count === 'string' ? parseInt(data.count) : data.count;
  if (typeof count !== 'number' || count <= 0) return;

  sqsClient.send(new SendMessageCommand({
    QueueUrl: process.env.SQS_URL,
    MessageBody: JSON.stringify({
      price,
      count,
      title: data.title,
      imageSrc: data.imageSrc,
      description: data.description,
    }),
  }));

};

const processProductsData = (getObjectResponse: GetObjectCommandOutput): Promise<unknown[]> => {
  const sqsClient = new SQSClient({ region: 'eu-west-1' });

  return new Promise((resolve) => {
    const results = [];
    getObjectResponse.Body
      ?.pipe(csv({ separator: ';' }))
      ?.on('data', (data) => { sendDataToQueue(sqsClient, data); })
      ?.on('end', () => { resolve(results) });
  });
}

export const importFileParser: S3Handler = async (event) => {
  try {
    const createdObjectKey = event.Records?.[0]?.s3?.object?.key;
    if (!createdObjectKey) return;

    const s3Client = new S3Client({ region: REGION });

    const getObjectResponse = await s3Client.send(new GetObjectCommand({
      Bucket: BUCKET_NAME,
      Key: createdObjectKey,
    }));

    await processProductsData(getObjectResponse);

    const fileName = getFileName(createdObjectKey);
    if (!fileName) return;

    await s3Client.send(new CopyObjectCommand({
      Bucket: BUCKET_NAME,
      Key: `parsed/${fileName}`,
      CopySource: `${BUCKET_NAME}/${createdObjectKey}`,
    }));

    await s3Client.send(new DeleteObjectCommand({
      Bucket: BUCKET_NAME,
      Key: createdObjectKey,
    }));
  } catch (e) {
    console.log('Internal server error appeared', e);
  }
};
