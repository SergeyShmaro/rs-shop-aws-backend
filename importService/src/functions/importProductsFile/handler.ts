import type { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
import { BUCKET_NAME, REGION } from 'src/constants/environment';

const isCsvFile = (fileName: string) => fileName.lastIndexOf('.csv') === (fileName.length - '.csv'.length);

export const importProductsFile: APIGatewayProxyHandlerV2<string> = async (event) => {
  try {
    const { name } = event.queryStringParameters;
    if (typeof name !== 'string' || name === '' || !isCsvFile(name)) {
      return { statusCode: 400, body: 'Provided file name is unacceptable' };
    }

    const s3Client = new S3Client({ region: REGION });

    const command = new PutObjectCommand({ Bucket: BUCKET_NAME, Key: `uploaded/${name}` });
    const signedUrl = await getSignedUrl(s3Client, command, { expiresIn: 3600, });
    return signedUrl;
  } catch (e) {
    console.log('Internal server error appeared', e);
    return { statusCode: 500, body: 'Internal server error appeared' };
  }
};
