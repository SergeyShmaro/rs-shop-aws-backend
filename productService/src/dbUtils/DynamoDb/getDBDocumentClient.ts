import { DynamoDB } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';

export default () => {
  const dbClient = new DynamoDB({ region: 'eu-west-1' });
  const documentClient = DynamoDBDocumentClient.from(dbClient);
  return documentClient;
};