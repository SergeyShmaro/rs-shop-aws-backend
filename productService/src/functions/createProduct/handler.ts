import { v4 as uuidV4 } from 'uuid';
import { TransactWriteCommand } from '@aws-sdk/lib-dynamodb';
import type { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import getDBDocumentClient from 'src/dbUtils/DynamoDb/getDBDocumentClient';
import { ProductWithStock } from 'src/types/apiTypes';

const isProvidedProductDataValid = (productData: any): productData is Omit<ProductWithStock, 'id'> => {
  if (typeof productData !== 'object') return false;
  if (typeof productData.title !== 'string') return false;
  if (typeof productData.price !== 'number') return false;
  if (typeof productData.count !== 'number') return false;
  if (productData.description !== undefined && typeof productData.description !== 'string') return false;
  if (productData.imageSrc !== undefined && typeof productData.imageSrc !== 'string') return false;
  return true;
};

export const createProduct: APIGatewayProxyHandlerV2<string> = async (event) => {
  console.log(`Start creating product with data: ${event.body}`, event);
  const productData = JSON.parse(event.body) ?? {};

  if (!isProvidedProductDataValid(productData)) {
    return { statusCode: 400, body: 'Provided product data is invalid' };
  }

  const { count, ...product } = productData;

  const newProductId = uuidV4();
  const productToInsert = { ...product, id: newProductId };
  const stockToInsert = { productId: newProductId, count };

  try {
    const documentClient = getDBDocumentClient();
    const transactionOutput = await documentClient.send(new TransactWriteCommand({
      TransactItems: [
        { Put: { Item: productToInsert, TableName: process.env.PRODUCTS_TABLE } },
        { Put: { Item: stockToInsert, TableName: process.env.STOCKS_TABLE } },
      ],
    }));

    const { httpStatusCode } = transactionOutput.$metadata;
    if (httpStatusCode !== 200) {
      throw new Error(`Request to DB returned ${httpStatusCode} status code`);
    };

    return { statusCode: 200, body: newProductId };
  } catch (e) {
    console.error('Error appeared during product creating', e);
    return { statusCode: 500, body: `Something went wrong during product retrieving` };
  }
};
