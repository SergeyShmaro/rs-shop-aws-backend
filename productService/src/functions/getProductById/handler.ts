import { QueryCommand } from '@aws-sdk/lib-dynamodb';
import type { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import getDBDocumentClient from 'src/dbUtils/DynamoDb/getDBDocumentClient';
import type { Product } from 'src/types/product';
import { ProductWithStock } from 'src/types/apiTypes';
import { Stock } from 'src/types/stock';
import { validate } from 'uuid';

export const getProductById: APIGatewayProxyHandlerV2<ProductWithStock> = async (event) => {
  const { productId } = event?.pathParameters ?? {};

  console.log(`Start retrieving product with id: ${productId}`, event);

  if (!validate(productId)) return { statusCode: 400, body: `Wrong productId parameter` };

  try {
    const documentClient = getDBDocumentClient();
    const products = await documentClient.send(new QueryCommand({
      TableName: process.env.PRODUCTS_TABLE,
      KeyConditionExpression: "id = :id",
      ExpressionAttributeValues: { ":id": productId },
    }));

    if (products.Items.length !== 1 || !products.Items[0]) {
      return { statusCode: 404, body: `Product with provided id was not found` };
    }

    const product = products.Items[0] as Product;

    const stocks = await documentClient.send(new QueryCommand({
      TableName: process.env.STOCKS_TABLE,
      KeyConditionExpression: "productId = :productId",
      ExpressionAttributeValues: { ":productId": productId },
    }));

    const stock = stocks.Items[0] as Stock;

    return { ...product, count: stock.count ?? 0 };
  } catch (e) {
    console.error('Error appeared during product retrieving', e);
    return { statusCode: 500, body: `Something went wrong during product retrieving` };
  }
};
