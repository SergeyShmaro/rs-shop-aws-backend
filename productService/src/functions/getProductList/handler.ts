import { ScanCommand } from '@aws-sdk/lib-dynamodb';
import type { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import getDBDocumentClient from 'src/dbUtils/DynamoDb/getDBDocumentClient';
import type { Product } from 'src/types/product';
import { ProductWithStock } from 'src/types/apiTypes';
import { Stock } from 'src/types/stock';

export const getProductList: APIGatewayProxyHandlerV2<ProductWithStock[]> = async (event) => {
  console.log('Start retrieving available products', event);
  try {
    const documentClient = getDBDocumentClient();
    const scanProductsOutput = await documentClient.send(new ScanCommand({
      TableName: process.env.PRODUCTS_TABLE,
    }));

    const products = (scanProductsOutput.Items as unknown) as Product[];
    const scanSrocksOutput = await documentClient.send(new ScanCommand({
      TableName: process.env.STOCKS_TABLE,
    }));

    const stocks = (scanSrocksOutput.Items as unknown) as Stock[];

    return products.map((product) => {
      const stock = stocks.find(({ productId }) => product.id === productId);
      return { ...product, count: stock.count };
    });
  } catch (e) {
    console.error('Error appeared during products retrieving', e);
    return {
      statusCode: 500,
      body: 'Error appeared during products retrieving',
    }
  }
};
