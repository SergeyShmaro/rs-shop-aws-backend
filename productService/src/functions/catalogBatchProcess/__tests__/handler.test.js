jest.mock('@aws-sdk/lib-dynamodb', () => ({ TransactWriteCommand: function () { } }));
jest.mock('@aws-sdk/client-sns', () => ({
  PublishCommand: jest.fn(),
  SNSClient: jest.fn().mockReturnValue({ send: jest.fn() }),
}));
import { SNSClient, PublishCommand } from '@aws-sdk/client-sns';
import { catalogBatchProcess } from '../handler'
import * as getDBDocumentClient from 'src/dbUtils/DynamoDb/getDBDocumentClient';

const testRecord1 = { title: 'testTitle1', price: 1, count: 1 };
const testRecord2 = { title: 'testTitle2', price: 2, count: 2 };
const wrongTestRecord1 = { title: 'testTitle3', price: -3, count: 3 };

const getRecords = (testRecords) => testRecords.map((record) => ({ body: JSON.stringify(record) }));

describe('catalogBatchProcess', () => {
  test('should create records inside Dynamo DB', async () => {
    const testRecords = [testRecord1, testRecord2];
    const eventStub = { Records: getRecords(testRecords) };

    const dynamoDbClientSendSpy = jest.fn().mockResolvedValue({ $metadata: { httpStatusCode: 200 } });
    jest.spyOn(getDBDocumentClient, 'default').mockImplementation(() => ({ send: dynamoDbClientSendSpy }));

    await catalogBatchProcess(eventStub);

    expect(dynamoDbClientSendSpy).toHaveBeenCalledTimes(testRecords.length);
    expect(dynamoDbClientSendSpy.mock.calls[0][0]).toMatchInlineSnapshot(`TransactWriteCommand {}`);
  });

  test('should create send messages to SNS service', async () => {
    const testRecords = [testRecord1, testRecord2];
    const eventStub = { Records: getRecords(testRecords) };

    const dynamoDbClientSendSpy = jest.fn().mockResolvedValue({ $metadata: { httpStatusCode: 200 } });
    jest.spyOn(getDBDocumentClient, 'default').mockImplementation(() => ({ send: dynamoDbClientSendSpy }));

    const snsSendSpy = jest.fn();
    SNSClient.mockReturnValue({ send: snsSendSpy });

    await catalogBatchProcess(eventStub);

    expect(snsSendSpy).toHaveBeenCalledTimes(testRecords.length);
    testRecords.forEach((record, index) => {
      expect(PublishCommand.mock.calls[index][0]).toMatchSnapshot();
    })
  });

  test('should create send error message to SNS service', async () => {
    const testRecords = [testRecord1, testRecord2, wrongTestRecord1];
    const eventStub = { Records: getRecords(testRecords) };

    const dynamoDbClientSendSpy = jest.fn().mockResolvedValue({ $metadata: { httpStatusCode: 500 } });
    jest.spyOn(getDBDocumentClient, 'default').mockImplementation(() => ({ send: dynamoDbClientSendSpy }));

    const snsSendSpy = jest.fn();
    SNSClient.mockReturnValue({ send: snsSendSpy });

    await catalogBatchProcess(eventStub);

    expect(snsSendSpy).toHaveBeenCalledTimes(1);
    expect(PublishCommand.mock.calls[0][0]).toMatchSnapshot();
  });
});
