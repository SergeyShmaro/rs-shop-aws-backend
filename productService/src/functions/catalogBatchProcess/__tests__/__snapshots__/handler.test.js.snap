// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catalogBatchProcess should create send error message to SNS service 1`] = `
{
  "Message": "
          Products failed to import: [{"title":"testTitle1","price":1,"count":1},{"title":"testTitle2","price":2,"count":2}];
          Products with wrong field values: [{"title":"testTitle3","price":-3,"count":3}];
        ",
  "Subject": "Products which was not imported",
  "TopicArn": undefined,
}
`;

exports[`catalogBatchProcess should create send messages to SNS service 1`] = `
{
  "Message": "Imported product data: {"title":"testTitle1","price":1,"count":1};",
  "MessageAttributes": {
    "price": {
      "DataType": "Number",
      "StringValue": "1",
    },
  },
  "Subject": "Product import",
  "TopicArn": undefined,
}
`;

exports[`catalogBatchProcess should create send messages to SNS service 2`] = `
{
  "Message": "Imported product data: {"title":"testTitle2","price":2,"count":2};",
  "MessageAttributes": {
    "price": {
      "DataType": "Number",
      "StringValue": "2",
    },
  },
  "Subject": "Product import",
  "TopicArn": undefined,
}
`;
